File: rice-cooker-golang/rice_cooker_model.go
package main

import (
	"fmt"
)

type RiceCookerModel struct {
	powerStatus bool
	waterLiters int
	riceGrams   int
	cooking     bool
}

func (rc *RiceCookerModel) Cook(minutes string) {
	if rc.powerStatus && rc.waterLiters > 0 && rc.riceGrams > 0 {
		fmt.Printf("Cuisson pendant %s minutes.\n", minutes)
		rc.cooking = true
	} else {
		fmt.Println("Impossible de cuire. Assurez-vous que le cuiseur de riz est allumé et contient de l'eau et du riz.")
	}
}

func (rc *RiceCookerModel) Boil(minutes string) {
	if rc.powerStatus && rc.waterLiters > 0 {
		fmt.Printf("Ébullition de l'eau pendant %s minutes.\n", minutes)
	} else {
		fmt.Println("Impossible de faire bouillir. Assurez-vous que le cuiseur de riz est allumé et contient de l'eau.")
	}
}

func (rc *RiceCookerModel) PowerOn() {
	rc.powerStatus = true
	fmt.Println("Le cuiseur de riz est allumé.")
}

func (rc *RiceCookerModel) PowerOff() {
	rc.powerStatus = false
	fmt.Println("Le cuiseur de riz est éteint.")
}

func (rc *RiceCookerModel) KeepWarm() {
	if rc.powerStatus {
		fmt.Println("Maintien au chaud du riz.")
	} else {
		fmt.Println("Impossible de maintenir au chaud. Allumez d'abord le cuiseur de riz.")
	}
}

func (rc *RiceCookerModel) StopWarm() {
	fmt.Println("Arrêt du maintien au chaud.")
}

func (rc *RiceCookerModel) GetStatus() string {
	if rc.powerStatus {
		return "Le cuiseur de riz est allumé."
	}
	return "Le cuiseur de riz est éteint."
}

func (rc *RiceCookerModel) AddWater(liters int) {
	rc.waterLiters += liters
	fmt.Printf("%d litres d'eau ajoutés.\n", liters)
}

func (rc *RiceCookerModel) AddRice(grams int) {
	rc.riceGrams += grams
	fmt.Printf("%d grammes de riz ajoutés.\n", grams)
}

func (rc *RiceCookerModel) IsCooking() bool {
	return rc.cooking
}

File: rice-cooker-on-scala/src/main/scala/RiceCookerModel.scala
package com.ricecooker

class RiceCookerModel extends TRiceCooker {
  private var powerStatus: Boolean = false
  private var waterLiters: Int = 0
  private var riceGrams: Int = 0
  private var cooking: Boolean = false

  override def cook(minutes: String): Unit = {
    if (powerStatus && waterLiters > 0 && riceGrams > 0) {
      println(s"Cuisson pendant $minutes minutes.")
      cooking = true
    } else {
      println("Impossible de cuisiner. Assurez-vous que le cuiseur de riz est allumé et contient de l'eau et du riz.")
    }
  }

  override def boil(minutes: String): Unit = {
    if (powerStatus && waterLiters > 0) {
      println(s"Ébullition d'eau pendant $minutes minutes.")
    } else {
      println("Impossible d'ébullition. Assurez-vous que le cuiseur de riz est allumé et contient de l'eau.")
    }
  }

  override def powerOn(): Unit = {
    powerStatus = true
    println("Le cuiseur de riz est allumé.")
  }

  override def powerOff(): Unit = {
    powerStatus = false
    println("Le cuiseur de riz est éteint.")
  }

  override def keepWarm(): Unit = {
    if (powerStatus) {
      println("Garder le riz au chaud.")
    } else {
      println("Impossible de garder au chaud. Allumez d'abord le cuiseur de riz.")
    }
  }

  override def stopWarm(): Unit = {
    println("Arrêt de la fonction de maintien au chaud.")
  }

  override def getStatus: String = {
    if (powerStatus) "Le cuiseur de riz est allumé." else "Le cuiseur de riz est éteint."
  }

  override def addWater(liters: Int): Unit = {
    waterLiters += liters
    println(s"$liters litres d'eau ajoutés.")
  }

  override def addRice(grams: Int): Unit = {
    riceGrams += grams
    println(s"$grams grammes de riz ajoutés.")
  }

  override def isCooking: Boolean = cooking
}

